File: klox/src/main/kotlin/dev/craigfurman/klox/Environment.kt
package dev.craigfurman.klox

class Environment(internal val enclosing: Environment? = null) {
    private val values = HashMap<String, Any?>()

    // TODO does this handle explicit/implicit nil binding?
    fun get(name: Token): Any? {
        if (values.containsKey(name.lexeme)) return values[name.lexeme]
        if (enclosing != null) return enclosing.get(name)
        throw RuntimeError(name, "Undefined variable '${name.lexeme}'.")
    }

    fun getAt(distance: Int, name: String) = ancestor(distance).values[name]

    fun define(name: String, value: Any?) {
        values[name] = value
    }

    fun assign(name: Token, value: Any?) {
        if (values.containsKey(name.lexeme)) {
            values[name.lexeme] = value
            return
        }
        if (enclosing != null) {
            enclosing.assign(name, value)
            return
        }
        throw RuntimeError(name, "Undefined variable '${name.lexeme}'.")
    }

    fun assignAt(distance: Int, name: Token, value: Any?) {
        ancestor(distance).values[name.lexeme] = value
    }

    private fun ancestor(distance: Int): Environment {
        var env = this
        repeat(distance) { env = env.enclosing!! }
        return env
    }
}

File: golox/lox/environment.go
package lox

import "fmt"

type Environment struct {
	values    map[string]any
	enclosing *Environment
}

func newEnvironment(enclosing *Environment) *Environment {
	return &Environment{enclosing: enclosing, values: map[string]any{}}
}

func (e *Environment) define(name string, value any) {
	e.values[name] = value
}

func (e *Environment) assign(name Token, value any) error {
	if _, ok := e.values[name.lexeme]; ok {
		e.values[name.lexeme] = value
		return nil
	}
	if e.enclosing != nil {
		return e.enclosing.assign(name, value)
	}
	return RuntimeError{token: name, message: fmt.Sprintf("Undefined variable '%s'.", name.lexeme)}
}

func (e *Environment) assignAt(distance int, name Token, value any) {
	e.ancestor(distance).values[name.lexeme] = value
}

func (e *Environment) get(name Token) (any, error) {
	if val, ok := e.values[name.lexeme]; ok {
		return val, nil
	}
	if e.enclosing != nil {
		return e.enclosing.get(name)
	}
	return nil, RuntimeError{token: name, message: fmt.Sprintf("Undefined variable '%s'.", name.lexeme)}
}

func (e *Environment) getAt(distance int, name string) any {
	return e.ancestor(distance).values[name]
}

func (e *Environment) ancestor(distance int) *Environment {
	env := e
	for i := 0; i < distance; i++ {
		env = env.enclosing
	}
	return env
}
